drivers/serial/uart_nrfx_uarte2.c

```diff
@@ -0,0 +1,1032 @@
+ * to use this approach because for constant input it can calculate nrf setting
+ * at compile time.
+ */
+#define NRF_BAUDRATE(baudrate) ((baudrate) == 300 ? 0x00014000 :\
+	(baudrate) == 600    ? 0x00027000 :			\
+	(baudrate) == 1200   ? NRF_UARTE_BAUDRATE_1200 :	\
+	(baudrate) == 2400   ? NRF_UARTE_BAUDRATE_2400 :	\
+	(baudrate) == 4800   ? NRF_UARTE_BAUDRATE_4800 :	\
+	(baudrate) == 9600   ? NRF_UARTE_BAUDRATE_9600 :	\
+	(baudrate) == 14400  ? NRF_UARTE_BAUDRATE_14400 :	\
+	(baudrate) == 19200  ? NRF_UARTE_BAUDRATE_19200 :	\
+	(baudrate) == 28800  ? NRF_UARTE_BAUDRATE_28800 :	\
+	(baudrate) == 31250  ? NRF_UARTE_BAUDRATE_31250 :	\
+	(baudrate) == 38400  ? NRF_UARTE_BAUDRATE_38400 :	\
+	(baudrate) == 56000  ? NRF_UARTE_BAUDRATE_56000 :	\
+	(baudrate) == 57600  ? NRF_UARTE_BAUDRATE_57600 :	\
+	(baudrate) == 76800  ? NRF_UARTE_BAUDRATE_76800 :	\
+	(baudrate) == 115200 ? NRF_UARTE_BAUDRATE_115200 :	\
+	(baudrate) == 230400 ? NRF_UARTE_BAUDRATE_230400 :	\
+	(baudrate) == 250000 ? NRF_UARTE_BAUDRATE_250000 :	\
+	(baudrate) == 460800 ? NRF_UARTE_BAUDRATE_460800 :	\
+	(baudrate) == 921600 ? NRF_UARTE_BAUDRATE_921600 :	\
+	(baudrate) == 1000000 ? NRF_UARTE_BAUDRATE_1000000 : 0)
+
+#define UARTE_DATA_FLAG_TRAMPOLINE	BIT(0)
+#define UARTE_DATA_FLAG_RX_ENABLED	BIT(1)
+
+struct uarte_async_data {
+	uart_callback_t user_callback;
+	void *user_data;
+
+	struct k_timer tx_timer;
+	struct k_timer rx_timer;
+
+	k_timeout_t rx_timeout;
+
+	/* Keeps the most recent error mask. */
+	uint32_t err;
+
+	uint8_t idle_cnt;
+};
+
+/* Device data structure */
+struct uarte_nrfx_data {
+	struct uart_async_to_irq_data *a2i_data;
+#if CONFIG_UART_USE_RUNTIME_CONFIGURE
+	struct uart_config uart_config;
+#endif
+	struct uarte_async_data *async;
+	atomic_t flags;
+	uint8_t  rx_byte;
+};
+BUILD_ASSERT(offsetof(struct uarte_nrfx_data, a2i_data) == 0);
+
+/* If set then pins are managed when going to low power mode. */
+#define UARTE_CFG_FLAG_GPIO_MGMT		BIT(0)
+
+/* If set then receiver is not used. */
+#define UARTE_CFG_FLAG_NO_RX			BIT(1)
+
+/* If set then instance is using interrupt driven API. */
+#define UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API	BIT(2)
+
+/**
+ * @brief Structure for UARTE configuration.
+ */
+struct uarte_nrfx_config {
+	const struct uart_async_to_irq_config *a2i_config;
+	nrfx_uarte_t nrfx_dev;
+	nrfx_uarte_config_t nrfx_config;
+	const struct pinctrl_dev_config *pcfg;
+	uint32_t flags;
+
+	LOG_INSTANCE_PTR_DECLARE(log);
+};
+BUILD_ASSERT(offsetof(struct uarte_nrfx_config, a2i_config) == 0);
+
+#define UARTE_ERROR_FROM_MASK(mask)					\
+	((mask) & NRF_UARTE_ERROR_OVERRUN_MASK ? UART_ERROR_OVERRUN	\
+	 : (mask) & NRF_UARTE_ERROR_PARITY_MASK ? UART_ERROR_PARITY	\
+	 : (mask) & NRF_UARTE_ERROR_FRAMING_MASK ? UART_ERROR_FRAMING	\
+	 : (mask) & NRF_UARTE_ERROR_BREAK_MASK ? UART_BREAK		\
+	 : 0)
+
+/* Determine if the device has interrupt driven API enabled. */
+#define IS_INT_DRIVEN_API(dev) \
+	(IS_ENABLED(CONFIG_UART_ASYNC_TO_INT_DRIVEN_API) && \
```

Comment:
In theory, `CONFIG_UART_ASYNC_TO_INT_DRIVEN_API` could be enabled for some other UART driver.
```suggestion
	(UARTE_ANY_INTERRUPT_DRIVEN && \
```